runtime/parachains/src/disputes.rs

```diff
@@ -0,0 +1,1963 @@
+			flags |= DisputeStateFlags::FOR_SUPERMAJORITY;
+		}
+
+		if state.validators_against.count_ones() >= supermajority_threshold {
+			flags |= DisputeStateFlags::AGAINST_SUPERMAJORITY;
+		}
+
+		flags
+	}
+}
+
+#[derive(PartialEq, RuntimeDebug)]
+enum SpamSlotChange {
+	Inc,
+	Dec,
+}
+
+struct ImportSummary<BlockNumber> {
+	// The new state, with all votes imported.
+	state: DisputeState<BlockNumber>,
+	// Changes to spam slots. Validator index paired with directional change.
+	spam_slot_changes: Vec<(ValidatorIndex, SpamSlotChange)>,
+	// Validators to slash for being (wrongly) on the AGAINST side.
+	slash_against: Vec<ValidatorIndex>,
+	// Validators to slash for being (wrongly) on the FOR side.
+	slash_for: Vec<ValidatorIndex>,
+	// New participants in the dispute.
+	new_participants: bitvec::vec::BitVec<BitOrderLsb0, u8>,
+	// Difference in state flags from previous.
+	new_flags: DisputeStateFlags,
+}
+
+#[derive(RuntimeDebug, PartialEq, Eq)]
+enum VoteImportError {
+	ValidatorIndexOutOfBounds,
+	DuplicateStatement,
+}
+
+impl<T: Config> From<VoteImportError> for Error<T> {
+	fn from(e: VoteImportError) -> Self {
+		match e {
+			VoteImportError::ValidatorIndexOutOfBounds => Error::<T>::ValidatorIndexOutOfBounds,
+			VoteImportError::DuplicateStatement => Error::<T>::DuplicateStatement,
+		}
+	}
+}
+
+struct DisputeStateImporter<BlockNumber> {
+	state: DisputeState<BlockNumber>,
+	now: BlockNumber,
+	new_participants: bitvec::vec::BitVec<BitOrderLsb0, u8>,
+	pre_flags: DisputeStateFlags,
+}
+
+impl<BlockNumber: Clone> DisputeStateImporter<BlockNumber> {
+	fn new(
+		state: DisputeState<BlockNumber>,
+		now: BlockNumber,
+	) -> Self {
+		let pre_flags = DisputeStateFlags::from_state(&state);
+		let new_participants = bitvec::bitvec![BitOrderLsb0, u8; 0; state.validators_for.len()];
+
+		DisputeStateImporter {
+			state,
+			now,
+			new_participants,
+			pre_flags,
+		}
+	}
+
+	fn import(&mut self, validator: ValidatorIndex, valid: bool)
+		-> Result<(), VoteImportError>
+	{
+		let (bits, other_bits) = if valid {
+			(&mut self.state.validators_for, &mut self.state.validators_against)
+		} else {
+			(&mut self.state.validators_against, &mut self.state.validators_for)
+		};
+
+		// out of bounds or already participated
+		match bits.get(validator.0 as usize).map(|b| *b) {
+			None => return Err(VoteImportError::ValidatorIndexOutOfBounds),
+			Some(true) => return Err(VoteImportError::DuplicateStatement),
+			Some(false) => {}
+		}
+
+		// inefficient, and just for extra sanity.
+		if validator.0 as usize >= self.new_participants.len() {
+			return Err(VoteImportError::ValidatorIndexOutOfBounds);
+		}
+
+		bits.set(validator.0 as usize, true);
+
+		// New participants tracks those which didn't appear on either
+		// side of the dispute until now. So we check the other side
+		// and checked the first side before.
+		if other_bits.get(validator.0 as usize).map_or(false, |b| !*b) {
```

Comment:
This would be an equivocation though right?

Reply:
Yeah.